relayer/message_relayer.go

```diff
@@ -548,27 +591,176 @@ func (r *messageRelayer) aggregateSignatures(signatureMap map[int]blsSignatureBu
 	return aggSig, vdrBitSet, nil
 }
 
+//
+// Database access
+//
+
+// Determines the height to process from. There are three cases:
+// 1) The database contains the latest processed block data for the chain
+//   - In this case, we return the maximum of the latest processed block and the configured processHistoricalBlocksFromHeight
+//
+// 2) The database has been configured for the chain, but does not contain the latest processed block data
+//   - In this case, we return the configured processHistoricalBlocksFromHeight
+//
+// 3) The database does not contain the any information for the chain.
+//   - In this case, we return the configured processHistoricalBlocksFromHeight if it is set, otherwise
+//     we return the chain head.
+func (r *messageRelayer) calculateStartingBlockHeight(processHistoricalBlocksFromHeight uint64) (uint64, error) {
```

Comment:
Think ahead to having a individual `MessageRelayers` (or `ApplicationRelayers`) for each source contract->destination contraction, are you thinking we should allow for specifying `processHistoricalBlocksFromHeight` in the configuration for each pair? I haven't thought about the pros/cons of that vs always using the DB value, or if it doesn't existing using the current block height as an initial starting point.

Reply:
I thought about this some, and where I landed was that we should keep `processHistoricalBlocksFromHeight` associated with the source chain, and not the application relayer. For one, the changes to implement it that way are very straightforward, and to do otherwise we'd either have to rearchitect the application to perform catch up at the application relayer level, or introduce a heuristic to determine the actual starting point for catch up. If the latter choice, we essentially arrive at what we have here, but with a config that doesn't as closely reflect the implemtation.

If there's a compelling use case for providing different heights to application relayers on startup, then we can reevaluate, but my opinion is that it's more trouble than it's worth atm.

Reply:
That makes sense to me. I definitely think we should have a single listener/filterer per source chain, which requires us to choose a single start block for all application relayers on that source chain. The implementation in this PR is reasonable in that it chooses the lowest checkpoint of all application relayers on that source chain, since there is no harm is re-scanning older messages for other application relayers. Given that though, I agree it makes sense to keep a single `processHistoricalBlocksFromHeight` for each source chain, rather than one per application relayer on that chain. 👍